File: problems/059_spiral_matrix_ii.go
/*
Given a positive integer n, generate an n x n matrix filled with elements from 1 to n2 in spiral order.

 

Example 1:


Input: n = 3
Output: [[1,2,3],[8,9,4],[7,6,5]]
Example 2:

Input: n = 1
Output: [[1]]
 

Constraints:

1 <= n <= 20
*/

func generateMatrix(n int) [][]int {
    matrix := make([][]int, n)
    for i := range matrix {
        matrix[i] = make([]int, n)
    }

    if n == 0 {
        return matrix
    }

    rowStart := 0
    rowEnd := n - 1
    colStart := 0
    colEnd := n - 1
    num := 1

    for rowStart <= rowEnd && colStart <= colEnd {
        for i := colStart; i <= colEnd; i++ {
            matrix[rowStart][i] = num
            num++
        }
        rowStart++

        for i := rowStart;i<=rowEnd; i++ {
            matrix[i][colEnd] = num
            num++
        }
        colEnd--
        for  i:= colEnd; i >= colStart; i-- {
            if rowStart <= rowEnd {
                matrix[rowEnd][i] = num
                num++
            }
        }
        rowEnd--
        for  i:= rowEnd; i >= rowStart; i-- {
            if colStart <= colEnd {
                matrix[i][colStart] = num
                num++
            }
        }
        colStart++
    }
    return matrix
}

File: problems/059_spiral_matrix_ii.rs
/*
Given a positive integer n, generate an n x n matrix filled with elements from 1 to n2 in spiral order.

 

Example 1:


Input: n = 3
Output: [[1,2,3],[8,9,4],[7,6,5]]
Example 2:

Input: n = 1
Output: [[1]]
 

Constraints:

1 <= n <= 20
*/

impl Solution {
    pub fn generate_matrix(n: i32) -> Vec<Vec<i32>> {
        let n = n as usize;
        let mut ret = vec![vec![0; n]; n];

        let mut num = 1;
        for i in 0..((n + 1) / 2) {
            for c in i..(n - 1 - i) {
                ret[i][c] = num;
                num += 1;
            }
            for r in i..(n - 1 - i) {
                ret[r][n - 1 - i] = num;
                num += 1;
            }
            for c in ((i + 1)..(n - i)).rev() {
                ret[n - 1 - i][c] = num;
                num += 1;
            }
            for r in ((i + 1)..(n - i)).rev() {
                ret[r][i] = num;
                num += 1;
            }
        }
        ret[n / 2][(n - 1) / 2] = (n * n) as i32;

        ret
    }
}

